features/src/autogluon/features/generators/text_ngram.py

```diff
@@ -67,6 +67,12 @@ def __init__(self, vectorizer=None, vectorizer_strategy='combined', max_memory_r
         self._feature_names_dict = dict()
 
     def _fit_transform(self, X: DataFrame, y: Series = None, problem_type: str = None, **kwargs) -> (DataFrame, dict):
+        for nlp_feature in self.features_in:
+            if hasattr(X[nlp_feature], 'cat'):
+                if '' not in X[nlp_feature].cat.categories:
+                    X[nlp_feature] = X[nlp_feature].cat.add_categories('').fillna('')
```

Comment:
You might be editing the outer context of the DataFrame. This would cause a lot of downstream issues. Please check that this is not editing the outer context.

If it is, the work-around is to perform a deepcopy on X before performing these operations.

Reply:
This is already a deepcopy made in the `general_data_processing`

Reply:
This is already a deepcopy made in `general_data_processing`

Reply:
`general_data_processing` is in the outer context relative to this operation, not inner context. For example, lets imagine we do the following:

Ngram preprocessing -> Cat preprocessing

In this situation, assuming we are editing outer context, X[cat_preprocessing_out] will now have a new category '', and NaN's are replaced by it because ngram preprocessing edits the outer context of X which alters the input of cat_preprocessing.

Swap order:

Cat preprocessing -> Ngram preprocessing

Now X[cat_preprocessing_out] will not have a new category '', and will still have NaN's.

If this is confusing, let me know and I can try to come up with a more concrete example.

Reply:
Actually they are not copied in feature generator, so could be a problem if the user uses feature generator directly

Reply:
Refer to benchmark script for more details, without deepcopy it alters outer context, plus I believe I found a faster alternative.